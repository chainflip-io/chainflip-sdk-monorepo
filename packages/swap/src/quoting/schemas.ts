import { z } from 'zod';
import { numericString, unsignedInteger } from '@/shared/parsers';

export const marketMakerResponseSchema = z.union([
  z
    .object({
      id: z.string(),
      intermediate_amount: numericString,
      output_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      intermediateAmount: rest.intermediate_amount,
      outputAmount: rest.output_amount,
    })),
  z
    .object({
      id: z.string(),
      output_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      outputAmount: rest.output_amount,
    })),
]);
export type MarketMakerQuote = z.infer<typeof marketMakerResponseSchema>;

export const swapRateResponseSchema = z
  .object({
    // TODO: simplify when we know how Rust `Option` is encoded
    intermediary: unsignedInteger.optional().nullable(),
    output: unsignedInteger,
  })
  .transform((rate) => ({
    intermediateAmount: rate.intermediary?.toString(),
    outputAmount: rate.output.toString(),
  }));
export type BrokerQuote = z.infer<typeof swapRateResponseSchema>;
