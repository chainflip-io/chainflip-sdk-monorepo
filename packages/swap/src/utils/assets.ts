import { SupportedAsset, supportedAsset } from '@/shared/enums';

export const isSupportedAsset = (value: string): value is SupportedAsset =>
  supportedAsset.safeParse(value).success;

export function assertSupportedAsset(
  value: string,
): asserts value is SupportedAsset {
  if (!isSupportedAsset(value)) {
    throw new Error(`received invalid asset "${value}"`);
  }
}

export const decimalPlaces: Record<SupportedAsset, number> = {
  DOT: 10,
  ETH: 18,
  FLIP: 18,
  USDC: 6,
  BTC: 8,
};
