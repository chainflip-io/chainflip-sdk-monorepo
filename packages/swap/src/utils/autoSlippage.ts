import BigNumber from 'bignumber.js';
import { assetConstants, chainConstants } from '@/shared/enums';
import { getDeployedLiquidity, getUndeployedLiquidity } from './pools';
import { getRequiredBlockConfirmations } from './rpc';
import { InternalAsset } from '../client';

const getLiquidityAutoSlippage = async (
  srcAsset: InternalAsset,
  destAsset: InternalAsset,
  amount: bigint,
) => {
  const deployedLiquidity = await getDeployedLiquidity(srcAsset, destAsset);
  const liquidityRatio = new BigNumber(amount.toString()).div(deployedLiquidity.toString());

  if (liquidityRatio.isGreaterThan(0.2)) {
    return liquidityRatio.multipliedBy(2);
  }
  return new BigNumber(0);
};

const getDepositTimeAutoSlippage = async (srcAsset: InternalAsset, isBoosted: boolean) => {
  const { chain } = assetConstants[srcAsset];
  const { blockTimeSeconds } = chainConstants[chain];
  const blockConfirmations = isBoosted ? 1 : (await getRequiredBlockConfirmations(srcAsset)) ?? 0;
  const depositTimeMinutes = (blockConfirmations * blockTimeSeconds) / 60;

  if (['Solana', 'Arbitrum'].includes(chain)) {
    return depositTimeMinutes * 0.01;
  }
  if (['Ethereum', 'Polkadot'].includes(chain)) {
    return depositTimeMinutes * 0.05;
  }
  return depositTimeMinutes * 0.1;
};

const getUndeployedLiquidityAutoSlippage = async (asset: InternalAsset, amount: bigint) => {
  const undeployedLiquidity = await getUndeployedLiquidity(asset);
  if (undeployedLiquidity >= amount) {
    return new BigNumber(-0.5);
  }

  const undeployedRatio = new BigNumber(undeployedLiquidity.toString()).div(amount.toString());
  if (undeployedRatio.gt(0.5)) {
    return new BigNumber(-undeployedRatio * 0.1);
  }

  return new BigNumber(0);
};

export const calculateRecommendedSlippage = async ({
  srcAsset,
  destAsset,
  intermediateAmount,
  egressAmount,
  boostFeeBps,
  dcaChunks,
}: {
  srcAsset: InternalAsset;
  destAsset: InternalAsset;
  intermediateAmount?: bigint;
  egressAmount: bigint;
  boostFeeBps?: number;
  dcaChunks: number;
}) => {
  const minSlippage = 0.1;
  const baseSlippage = 1;

  const assetTo = srcAsset === 'Usdc' || destAsset === 'Usdc' ? destAsset : 'Usdc';

  const [deployedLiquiditySlippage, timeSlippage, undeployedLiquiditySlippage] = await Promise.all([
    await getLiquidityAutoSlippage(
      srcAsset,
      assetTo,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
    await getDepositTimeAutoSlippage(srcAsset, Boolean(boostFeeBps && boostFeeBps > 0)),
    await getUndeployedLiquidityAutoSlippage(
      intermediateAmount ? 'Usdc' : destAsset,
      (intermediateAmount ?? egressAmount) * BigInt(dcaChunks),
    ),
  ]);

  const calculatedSlippage = deployedLiquiditySlippage
    .plus(timeSlippage)
    .plus(undeployedLiquiditySlippage)
    .plus(baseSlippage);

  return Math.max(minSlippage, Math.round(calculatedSlippage.toNumber() * 100) / 100);
};
