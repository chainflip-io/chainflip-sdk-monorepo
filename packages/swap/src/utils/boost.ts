import { ONE_IN_PIP, bigintMin, getPipAmountFromAmount } from '@/shared/functions';
import { BoostPoolsDepth } from '@/shared/rpc';

export const getBoostFeeBpsForAmount = async ({
  amount,
  assetBoostPoolsDepth,
}: {
  amount: bigint;
  assetBoostPoolsDepth: BoostPoolsDepth;
}): Promise<number | undefined> => {
  assetBoostPoolsDepth.sort((a, b) => (a.tier < b.tier ? -1 : 1));

  let _amount = amount;
  let feeAmount = 0n;

  for (const poolDepth of assetBoostPoolsDepth) {
    const poolAvailableAmount = poolDepth.availableAmount;

    const amountToBeUsedFromPool = bigintMin(_amount, poolAvailableAmount);
    feeAmount += getPipAmountFromAmount(amountToBeUsedFromPool, poolDepth.tier);
    _amount -= amountToBeUsedFromPool;
  }

  // Not enough liquidity in the boost pools
  if (_amount > 0) return undefined;

  return Number((feeAmount * BigInt(ONE_IN_PIP)) / BigInt(amount));
};
