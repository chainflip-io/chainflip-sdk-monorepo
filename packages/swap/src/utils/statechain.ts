import { z } from 'zod';
import { SupportedAsset, supportedAsset } from '@/shared/enums';
import { numericString } from '@/shared/parsers';
import { QuoteResponse, QuoteQueryParams } from '../schemas';
import { memoize } from './function';
import RpcClient from './RpcClient';
import { transformAsset } from './string';

const requestValidators = {
  swap_rate: z.tuple([
    supportedAsset.transform(transformAsset),
    supportedAsset.transform(transformAsset),
    numericString,
  ]),
};

// parse hex encoding or decimal encoding into decimal encoding
const assetAmount = z.string().transform((v) => BigInt(v).toString());

const responseValidators = {
  swap_rate: z.object({
    // TODO: simplify when we know how Rust `Option` is encoded
    intermediary: assetAmount.optional().nullable(),
    output: assetAmount,
  }),
};

const initializeClient = memoize(async () => {
  const rpcClient = await new RpcClient(
    process.env.RPC_NODE_WSS_URL as string,
    requestValidators,
    responseValidators,
    'cf',
  ).connect();

  return rpcClient;
});

const getSwapAmount = async (
  srcAsset: SupportedAsset,
  destAsset: SupportedAsset,
  amount: string,
): Promise<z.infer<(typeof responseValidators)['swap_rate']>> => {
  const client = await initializeClient();

  return client.sendRequest('swap_rate', srcAsset, destAsset, amount);
};

export const getBrokerQuote = async (
  { srcAsset, destAsset, amount }: QuoteQueryParams,
  id: string,
): Promise<QuoteResponse> => {
  const { intermediary, output } = await getSwapAmount(
    srcAsset,
    destAsset,
    amount,
  );

  return {
    id,
    intermediateAmount: intermediary ?? undefined,
    egressAmount: output,
  };
};
