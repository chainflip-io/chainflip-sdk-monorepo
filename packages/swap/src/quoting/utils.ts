import assert from 'assert';
import BigNumber from 'bignumber.js';
import { filter, Observable, Subscription } from 'rxjs';
import { Side, type SwapInput } from '@/amm-addon';
import { InternalAsset, getAssetAndChain } from '@/shared/enums';
import { MarketMakerQuote, Leg as MarketMakerLeg } from './schemas';
import env from '../config/env';
import { getAssetPrice } from '../pricing';
import { assertUnreachable } from '../utils/function';
import { SwapRateArgs } from '../utils/statechain';

export type QuoteType = 'pool' | 'market_maker';

export const collectMakerQuotes = (
  requestId: string,
  expectedQuotes: number,
  quotes$: Observable<{ marketMaker: string; quote: MarketMakerQuote }>,
): Promise<MarketMakerQuote[]> => {
  if (expectedQuotes === 0) return Promise.resolve([]);

  const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();

  return new Promise((resolve) => {
    let sub: Subscription;

    const complete = () => {
      sub.unsubscribe();
      resolve([...clientsReceivedQuotes.values()]);
    };

    sub = quotes$
      .pipe(filter(({ quote }) => quote.request_id === requestId))
      .subscribe(({ marketMaker, quote }) => {
        clientsReceivedQuotes.set(marketMaker, quote);
        if (clientsReceivedQuotes.size === expectedQuotes) complete();
      });

    setTimeout(complete, env.QUOTE_TIMEOUT);
  });
};

export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
  const price = await getAssetPrice(asset);

  if (typeof price !== 'number') return null;

  return BigInt(new BigNumber(amount).times(price).toFixed(0));
};

type BaseAsset = Exclude<InternalAsset, 'Usdc'>;

export class Leg {
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint): Leg;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null): Leg | null;
  static of(baseAsset: InternalAsset, quoteAsset: InternalAsset, amount: bigint | null) {
    if (amount === null) return null;
    assert(baseAsset !== quoteAsset, 'baseAsset and quoteAsset must be different');
    assert(baseAsset === 'Usdc' || quoteAsset === 'Usdc', 'one of the assets must be USDC');
    return new Leg(baseAsset, quoteAsset, amount);
  }

  private constructor(
    private readonly baseAsset: InternalAsset,
    private readonly quoteAsset: InternalAsset,
    public amount: bigint,
  ) {}

  toPoolJSON(): SwapRateArgs {
    return {
      amount: this.amount,
      srcAsset: this.baseAsset,
      destAsset: this.quoteAsset,
    };
  }

  private getSide(): 'BUY' | 'SELL' {
    return this.quoteAsset !== 'Usdc' ? 'BUY' : 'SELL';
  }

  toMarketMakerJSON(): MarketMakerLeg {
    const side = this.getSide();
    let normalizedBaseAsset: BaseAsset;
    const normalizedQuoteAsset = 'Usdc';

    if (this.quoteAsset !== 'Usdc') {
      normalizedBaseAsset = this.quoteAsset;
    } else if (this.baseAsset !== 'Usdc') {
      normalizedBaseAsset = this.baseAsset;
    } else {
      return assertUnreachable('invalid leg');
    }

    return {
      base_asset: getAssetAndChain(normalizedBaseAsset),
      quote_asset: getAssetAndChain(normalizedQuoteAsset),
      amount: this.amount.toString(),
      side,
    };
  }

  toSwapInput(): SwapInput {
    return {
      side: this.getSide() === 'BUY' ? Side.Buy : Side.Sell,
      amount: this.amount,
      limitOrders: [],
    };
  }
}
