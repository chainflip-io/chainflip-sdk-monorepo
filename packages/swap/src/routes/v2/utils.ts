import { getAssetAndChain } from '@/shared/enums';
import { assertUnreachable } from '@/shared/functions';
import { StateV2 } from './swap';
import prisma, {
  Prisma,
  Swap,
  Egress,
  Broadcast,
  SwapFee,
  IgnoredEgress,
  FailedSwap,
} from '../../client';
import { getPendingBroadcast } from '../../ingress-egress-tracking';
import { failedSwapMessage } from '../../utils/swap';
import { FailureMode } from '../swap';

export const depositChannelInclude = {
  failedBoosts: true,
  failedSwaps: true,
  affiliates: {
    select: {
      account: true,
      commissionBps: true,
    },
  },
} as const;

export const swapRequestInclude = {
  swaps: { include: { fees: true }, orderBy: { nativeId: 'desc' } },
  egress: { include: { broadcast: true } },
  refundEgress: { include: { broadcast: true } },
  fees: true,
  ignoredEgresses: true,
  swapDepositChannel: {
    include: depositChannelInclude,
  },
} as const;

export const getSwapFields = (swap: Swap & { fees: SwapFee[] }) => ({
  swapInputAmount: swap.swapInputAmount.toString(),
  swapOutputAmount: swap.swapOutputAmount?.toString(),
  scheduledAt: swap.swapScheduledAt.valueOf(),
  scheduledBlockIndex: swap.swapScheduledBlockIndex ?? undefined,
  executedAt: swap.swapExecutedAt?.valueOf(),
  executedBlockIndex: swap.swapExecutedBlockIndex ?? undefined,
  retryCount: swap.retryCount,
  latestSwapRescheduledAt: swap.latestSwapRescheduledAt?.valueOf(),
  latestSwapRescheduledBlockIndex: swap.latestSwapRescheduledBlockIndex ?? undefined,
  fees: swap.fees.map((fee) => ({
    type: fee.type,
    ...getAssetAndChain(fee.asset),
    amount: fee.amount.toString(),
  })),
});

export const getDepositIgnoredFailedState = (failedSwap: FailedSwap) => ({
  failedAt: failedSwap.failedAt.valueOf(),
  failedAtBlockIndex: failedSwap.failedBlockIndex,
  mode: FailureMode.IngressIgnored,
  reason: {
    name: failedSwap.reason,
    message: failedSwapMessage[failedSwap.reason],
  },
});

export const getEgressFailureState = async (
  ignoredEgress: IgnoredEgress | undefined,
  broadcast: Broadcast | null | undefined,
  type: IgnoredEgress['type'] | undefined,
) => {
  const abortedBroadcast = broadcast?.abortedAt?.valueOf();
  const failedAt = abortedBroadcast ?? ignoredEgress?.ignoredAt?.valueOf();
  const failedAtBlockIndex = broadcast?.abortedBlockIndex ?? ignoredEgress?.ignoredBlockIndex;

  let error;
  let failureMode;

  if (failedAt && failedAtBlockIndex) {
    error = {
      name: 'Unknown',
      message: 'An unknown error occurred',
    };
    if (ignoredEgress) {
      switch (ignoredEgress.type) {
        case 'REFUND':
          failureMode = FailureMode.RefundEgressIgnored;
          break;
        case 'SWAP':
          failureMode = FailureMode.EgressIgnored;
          break;
        default:
          assertUnreachable(ignoredEgress.type);
      }
      const [stateChainError] = await Promise.all([
        prisma.stateChainError.findUniqueOrThrow({
          where: { id: ignoredEgress.stateChainErrorId },
        }),
      ]);
      error = {
        name: stateChainError.name,
        message: stateChainError.docs,
      };
    } else if (abortedBroadcast) {
      const message =
        type === 'SWAP' ? 'The swap broadcast was aborted' : 'The refund broadcast was aborted';
      failureMode = FailureMode.BroadcastAborted;
      error = {
        name: 'BroadcastAborted',
        message,
      };
    }
    return {
      failedAt,
      failedAtBlockIndex,
      mode: failureMode,
      reason: error,
    };
  }
  return undefined;
};

export const getEgressStatusFields = async (
  egress: Egress | null | undefined,
  broadcast: Broadcast | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  type: IgnoredEgress['type'] | undefined,
  egressTrackerTxRef: string | null | undefined,
) => {
  const ignoredEgress = ignoredEgresses?.find((e) => e.type === type);
  const failureState = await getEgressFailureState(ignoredEgress, broadcast, type);
  return {
    ...(egress && {
      outputAmount: egress.amount?.toFixed(),
      scheduledAt: egress.scheduledAt?.valueOf(),
      scheduledBlockIndex: egress.scheduledBlockIndex ?? undefined,
    }),
    ...(broadcast && {
      sentAt: broadcast?.succeededAt?.valueOf(),
      sentAtBlockIndex: broadcast?.succeededBlockIndex ?? undefined,
      sentTxRef: broadcast?.transactionRef ?? egressTrackerTxRef,
    }),
    ...(failureState && { failure: failureState }),
    ...(ignoredEgress && { ignoredAmount: ignoredEgress.amount?.toFixed() }),
  };
};

export const getSwapState = async (
  failedSwap: FailedSwap | null | undefined,
  ignoredEgresses: IgnoredEgress[] | undefined,
  swapRequest:
    | Prisma.SwapRequestGetPayload<{
        include: typeof swapRequestInclude;
      }>
    | undefined
    | null,
) => {
  let state: StateV2 | undefined;
  let swapEgressTrackerTxRef: string | null | undefined;
  let refundEgressTrackerTxRef: string | null | undefined;
  const swapEgress = swapRequest?.egress;
  const refundEgress = swapRequest?.refundEgress;
  const egress = swapEgress ?? refundEgress;

  if (
    failedSwap ||
    (ignoredEgresses && ignoredEgresses.length > 0) ||
    egress?.broadcast?.abortedAt
  ) {
    state = StateV2.Failed;
  } else if (egress?.broadcast?.succeededAt) {
    state = StateV2.Completed;
  } else if (egress?.broadcast) {
    if (swapEgress?.broadcast) {
      state = StateV2.Sending;
      const pendingSwapBroadcast = await getPendingBroadcast(swapEgress.broadcast);
      if (pendingSwapBroadcast) {
        state = StateV2.Sent;
        swapEgressTrackerTxRef = pendingSwapBroadcast.tx_ref;
      }
    }
    if (refundEgress?.broadcast) {
      const pendingRefundBroadcast = await getPendingBroadcast(refundEgress.broadcast);
      if (pendingRefundBroadcast) {
        state = StateV2.Sent;
        refundEgressTrackerTxRef = pendingRefundBroadcast.tx_ref;
      }
    }
  } else if (egress) {
    state = StateV2.Sending;
  } else if (swapRequest?.swaps.some((s) => s.swapScheduledAt)) {
    state = StateV2.Swapping;
  } else {
    state = StateV2.Receiving;
  }

  return {
    state,
    swapEgressTrackerTxRef,
    refundEgressTrackerTxRef,
  };
};
