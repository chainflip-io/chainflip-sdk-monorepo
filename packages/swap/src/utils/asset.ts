import { ChainflipAsset } from '@chainflip/utils/chainflip';

export const isAssetHub = (asset: ChainflipAsset): asset is 'HubDot' | 'HubUsdc' | 'HubUsdt' =>
  asset === 'HubDot' || asset === 'HubUsdc' || asset === 'HubUsdt';
