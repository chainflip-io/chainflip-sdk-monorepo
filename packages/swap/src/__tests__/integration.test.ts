import { hexEncodeNumber } from '@chainflip/utils/number';
import axios from 'axios';
import { spawn, ChildProcessWithoutNullStreams, exec } from 'child_process';
import * as crypto from 'crypto';
import { on, once } from 'events';
import { AddressInfo } from 'net';
import * as path from 'path';
import { Observable, filter, firstValueFrom, from, map, shareReplay, timeout } from 'rxjs';
import { promisify } from 'util';
import { Assets, Chains } from '@/shared/enums';
import { QuoteQueryParams } from '@/shared/schemas';
import {
  boostPoolsDepth,
  cfAccountInfo,
  cfPoolDepth,
  environment,
  mockRpcResponse,
} from '@/shared/tests/fixtures';
import prisma from '../client';
import app from '../server';
import { getSwapRateV3 } from '../utils/statechain';

const execAsync = promisify(exec);

jest.mock('../pricing');

jest.mock('../utils/statechain', () => ({
  getSwapRateV3: jest.fn().mockImplementation(() => Promise.reject(new Error('unexpected call'))),
}));

const generateKeyPairAsync = promisify(crypto.generateKeyPair);

describe('python integration test', () => {
  jest.setTimeout(10000);

  let privateKey: string;
  const accountId = 'web_team_whales';
  let server: typeof app;
  let child: ChildProcessWithoutNullStreams;
  let stdout$: Observable<string>;
  let serverUrl: string;

  beforeAll(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE public."Pool" CASCADE`;
    await prisma.pool.createMany({
      data: [
        {
          baseAsset: 'Flip',
          quoteAsset: 'Usdc',
          liquidityFeeHundredthPips: 1000,
        },
        {
          baseAsset: 'Eth',
          quoteAsset: 'Usdc',
          liquidityFeeHundredthPips: 2000,
        },
      ],
    });
  });

  beforeEach(async () => {
    mockRpcResponse((url, data) => {
      if (data.method === 'cf_environment') {
        return Promise.resolve({
          data: environment({
            maxSwapAmount: null,
            ingressFee: hexEncodeNumber(2000000),
            egressFee: hexEncodeNumber(50000),
          }),
        });
      }

      if (data.method === 'cf_boost_pools_depth') {
        return Promise.resolve({ data: boostPoolsDepth([]) });
      }

      if (data.method === 'cf_pool_depth') {
        return Promise.resolve({
          data: cfPoolDepth(),
        });
      }

      if (data.method === 'cf_accounts') {
        return Promise.resolve({
          data: {
            id: 1,
            jsonrpc: '2.0',
            result: [
              ['cFMYYJ9F1r1pRo3NBbnQDVRVRwY9tYem39gcfKZddPjvfsFfH', 'Chainflip Testnet Broker 2'],
            ],
          },
        });
      }

      if (data.method === 'cf_account_info') {
        return Promise.resolve({
          data: cfAccountInfo(),
        });
      }

      throw new Error(`unexpected axios call to ${url}: ${JSON.stringify(data)}`);
    });

    await prisma.$queryRaw`TRUNCATE TABLE private."MarketMaker" CASCADE`;
    const pair = await generateKeyPairAsync('ed25519');
    privateKey = pair.privateKey.export({ type: 'pkcs8', format: 'pem' }).toString();
    await prisma.marketMaker.create({
      data: {
        name: accountId,
        publicKey: pair.publicKey.export({ type: 'spki', format: 'pem' }).toString(),
      },
    });
    server = app.listen(0);
    serverUrl = `http://127.0.0.1:${(server.address() as AddressInfo).port}`;

    const exeName = await Promise.any([
      execAsync('python --version').then(() => 'python'),
      execAsync('python3 --version').then(() => 'python3'),
    ]);

    child = spawn(exeName, [
      path.join(__dirname, '..', '..', 'python-client', 'mock.py'),
      '--private-key',
      privateKey,
      '--account-id',
      accountId,
      '--url',
      serverUrl,
    ]);

    stdout$ = from(on(child.stdout, 'data')).pipe(
      map((buffer) => buffer.toString().trim()),
      shareReplay(),
    );
  });

  afterEach(async () => {
    if (child.exitCode === null) {
      child.kill('SIGINT');
      await once(child, 'exit');
    }
    await promisify(server.close).bind(server)();
  });

  const expectMesage = async (message: string) => {
    if (child.exitCode !== null) {
      throw new Error('child process exited unexpectedly');
    }

    await expect(
      Promise.race([
        firstValueFrom(
          stdout$.pipe(
            filter((msg) => msg === message),
            timeout(10000),
          ),
        ),
        once(child, 'close').then(() => Promise.reject(Error('child process exited unexpectedly'))),
      ]),
    ).resolves.toBe(message);
  };

  it('replies to a quote request', async () => {
    await expectMesage('connected');

    const query = {
      srcAsset: Assets.FLIP,
      srcChain: Chains.Ethereum,
      destAsset: Assets.USDC,
      destChain: Chains.Ethereum,
      amount: '1000000000000000000',
    } as QuoteQueryParams;
    const params = new URLSearchParams(query as Record<string, any>);

    jest.mocked(getSwapRateV3).mockResolvedValueOnce({
      ingressFee: { amount: 2000000n, chain: 'Ethereum', asset: 'FLIP' },
      networkFee: { amount: 998900109987003n, chain: 'Ethereum', asset: 'USDC' },
      egressFee: { amount: 50000n, chain: 'Ethereum', asset: 'USDC' },
      intermediateAmount: 2000000000n,
      egressAmount: 997901209876966295n,
      brokerFee: {
        chain: 'Ethereum',
        asset: 'USDC',
        amount: 0n,
      },
    });

    const response = await axios.get(`${serverUrl}/quote?${params.toString()}`);

    expect(await response.data).toMatchSnapshot();
    expect(jest.mocked(getSwapRateV3).mock.calls).toMatchSnapshot();
  });

  it('replies to a quote request with v2 endpoint', async () => {
    await expectMesage('connected');

    const query = {
      srcAsset: Assets.FLIP,
      srcChain: Chains.Ethereum,
      destAsset: Assets.USDC,
      destChain: Chains.Ethereum,
      amount: '1000000000000000000',
    } as QuoteQueryParams;
    const params = new URLSearchParams(query as Record<string, any>);

    jest.mocked(getSwapRateV3).mockResolvedValueOnce({
      ingressFee: { amount: 2000000n, chain: 'Ethereum', asset: 'FLIP' },
      networkFee: { amount: 998900109987003n, chain: 'Ethereum', asset: 'USDC' },
      egressFee: { amount: 50000n, chain: 'Ethereum', asset: 'USDC' },
      intermediateAmount: 2000000000n,
      egressAmount: 997901209876966295n,
      brokerFee: {
        chain: 'Ethereum',
        asset: 'USDC',
        amount: 0n,
      },
    });

    const response = await axios.get(`${serverUrl}/v2/quote?${params.toString()}`);

    expect(await response.data).toMatchSnapshot();
    expect(jest.mocked(getSwapRateV3).mock.calls).toMatchSnapshot();
  });
});
