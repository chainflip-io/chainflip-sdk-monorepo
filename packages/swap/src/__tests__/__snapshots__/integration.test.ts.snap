// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`python integration test > replies to a quote request 1`] = `
{
  "depositAmount": "1000000000000000000",
  "destAsset": {
    "asset": "USDC",
    "chain": "Ethereum",
  },
  "egressAmount": "997901209876966295",
  "estimatedDurationSeconds": 144,
  "estimatedDurationsSeconds": {
    "deposit": 30,
    "egress": 102,
    "swap": 12,
  },
  "estimatedPrice": "997901209879.01209741975802419484",
  "includedFees": [
    {
      "amount": "2000000",
      "asset": "FLIP",
      "chain": "Ethereum",
      "type": "INGRESS",
    },
    {
      "amount": "998900109987003",
      "asset": "USDC",
      "chain": "Ethereum",
      "type": "NETWORK",
    },
    {
      "amount": "50000",
      "asset": "USDC",
      "chain": "Ethereum",
      "type": "EGRESS",
    },
  ],
  "intermediateAmount": "2000000000",
  "isVaultSwap": false,
  "lowLiquidityWarning": false,
  "poolInfo": [
    {
      "baseAsset": {
        "asset": "FLIP",
        "chain": "Ethereum",
      },
      "fee": {
        "amount": "999999999998000",
        "asset": "FLIP",
        "chain": "Ethereum",
      },
      "quoteAsset": {
        "asset": "USDC",
        "chain": "Ethereum",
      },
    },
  ],
  "recommendedSlippageTolerancePercent": 5,
  "srcAsset": {
    "asset": "FLIP",
    "chain": "Ethereum",
  },
  "type": "REGULAR",
}
`;

exports[`python integration test > replies to a quote request 2`] = `
[
  [
    {
      "brokerCommissionBps": undefined,
      "ccmParams": undefined,
      "dcaParams": undefined,
      "depositAmount": 1000000000000000000n,
      "destAsset": "Usdc",
      "excludeFees": [],
      "limitOrders": [
        {
          "LimitOrder": {
            "base_asset": {
              "asset": "FLIP",
              "chain": "Ethereum",
            },
            "quote_asset": {
              "asset": "USDC",
              "chain": "Ethereum",
            },
            "sell_amount": "0xde0b6b3a7640000",
            "side": "buy",
            "tick": -1,
          },
        },
      ],
      "srcAsset": "Flip",
    },
  ],
]
`;

exports[`python integration test > replies to a quote request with v2 endpoint 1`] = `
[
  {
    "depositAmount": "1000000000000000000",
    "destAsset": {
      "asset": "USDC",
      "chain": "Ethereum",
    },
    "egressAmount": "997901209876966295",
    "estimatedDurationSeconds": 144,
    "estimatedDurationsSeconds": {
      "deposit": 30,
      "egress": 102,
      "swap": 12,
    },
    "estimatedPrice": "997901209879.01209741975802419484",
    "includedFees": [
      {
        "amount": "2000000",
        "asset": "FLIP",
        "chain": "Ethereum",
        "type": "INGRESS",
      },
      {
        "amount": "998900109987003",
        "asset": "USDC",
        "chain": "Ethereum",
        "type": "NETWORK",
      },
      {
        "amount": "50000",
        "asset": "USDC",
        "chain": "Ethereum",
        "type": "EGRESS",
      },
    ],
    "intermediateAmount": "2000000000",
    "isVaultSwap": false,
    "lowLiquidityWarning": false,
    "poolInfo": [
      {
        "baseAsset": {
          "asset": "FLIP",
          "chain": "Ethereum",
        },
        "fee": {
          "amount": "999999999998000",
          "asset": "FLIP",
          "chain": "Ethereum",
        },
        "quoteAsset": {
          "asset": "USDC",
          "chain": "Ethereum",
        },
      },
    ],
    "recommendedSlippageTolerancePercent": 5,
    "srcAsset": {
      "asset": "FLIP",
      "chain": "Ethereum",
    },
    "type": "REGULAR",
  },
]
`;

exports[`python integration test > replies to a quote request with v2 endpoint 2`] = `
[
  [
    {
      "brokerCommissionBps": undefined,
      "ccmParams": undefined,
      "dcaParams": undefined,
      "depositAmount": 1000000000000000000n,
      "destAsset": "Usdc",
      "excludeFees": [],
      "limitOrders": [
        {
          "LimitOrder": {
            "base_asset": {
              "asset": "FLIP",
              "chain": "Ethereum",
            },
            "quote_asset": {
              "asset": "USDC",
              "chain": "Ethereum",
            },
            "sell_amount": "0xde0b6b3a7640000",
            "side": "buy",
            "tick": -1,
          },
        },
      ],
      "srcAsset": "Flip",
    },
  ],
]
`;
